scalyr_agent/monitor_utils/k8s.py

```diff
@@ -2400,6 +2401,138 @@ def query_stats(self):
+        self, k8s_log_configs, logger, rename_no_original, parse_format="json"
+    ):
+        """
+        @param k8s_log_configs: The config snippets from the configuration
+        @param logger: A scalyr logger
+        @param rename_no_original: A bool, used to prevent the original log file name from being added to the attributes.
+        @param parse_format: The parse format of this log config
+        """
+        self.__k8s_log_configs = k8s_log_configs
+        self._logger = logger
+        self.__rename_no_original = rename_no_original
+        self.__parse_format = parse_format
+
+    def _check_match(self, element, name, value, glob):
+        """
+        Checks to see if we have a match against the glob for a certain value
+        @param element: The index number of the element in the k8s_config list
+        @param name: A string containing the name of the field we are evaluating (used for debug log purposes)
+        @param value: The value of the field to evaluate the glob against
+        @param glob: A string containing a glob to evaluate
+        """
+        result = False
+        if glob is not None and value is not None:
+            # ignore this config if value doesn't match the glob
+            if fnmatch.fnmatch(value, glob):
+                result = True
+            else:
+                self._logger.log(
+                    scalyr_logging.DEBUG_LEVEL_2,
+                    "Ignoring k8s_log item %d because %s '%s' doesn't match '%s'"
+                    % (element, name, value, glob,),
+                )
+        return result
+
+    def get_log_config(
+        self, info, k8s_info, container_attributes, parser, rename_logfile,
+    ):
+        """
+        Creates a log_config from various attributes and then applies any `k8s_logs` configs that
+        might apply to this log
+        @param info: A dict containing docker information about the container we are creating a config for
+        @param k8s_info: A dict containing k8s information about hte container we are creating a config for
+        @param container_attributes: A set of attributes to add to the log config of this container
+        @param parser: A string containing the name of the parser to use for this log config
+        @param rename_logfile: A string containing the name to use for the renamed log file
+        @return: A dict containing a log_config, or None if we couldn't create a valid config (i.e. log_path was empty)
+        """
+
+        # Make sure we have a log_path for the log config
+        path = info.get("log_path", None)
+        if not path:
+            return None
+
+        # Build up the default config that we will use
+        result = {
+            "parser": parser,
+            "path": path,
+            "parse_format": self.__parse_format,
+            "attributes": container_attributes,
+            "rename_logfile": rename_logfile,
+            "rename_no_original": self.__rename_no_original,
```

Comment:
If I understand this correctly, `parser` could either be `docker` or a value from the pod label.  rename_logfile will always be a fixed format (a combination of CRI and the container name) just varies based on the actual container specifics.  And, attributes is always just the K8s attributes that we add in.

Is that correct?

I wonder if we should respect the parser from the k8s pods if there is a configuration that matches this pod.  I worry that it might be confusing to the customer if we try to merge in information from the various configuration mechanisms we have.  Part of me wants to say "pick one and stick with it" because otherwise it is confusing.

What do you think?

Reply:
Part of the reason I'm pushing on these questions is that I don't think it's a good user experience if overriding some options for the log file forces you change something you may not want to change.

Take the `rename_logfile` config option.  By default, k8s logs are renamed to "/${CONTAINER_RT}/${CONTAINER_NAME}".  However, if you override the log config for particular log to set a parser, there's no way for you to also get the same "/${CONTAINER_RT}/${CONTAINER_NAME}" name.  That's because you have to specify some value for the `rename_logfile` config option.. but you can't specify one that will provide the exact same behavior.  Now, we can take the approach you took and just have that one be the default if you don't specify `rename_logfile`.  But, that seems like almost a hidden behavior.

Instead, we should just officially same the default for the K8s log file rename is "/${CONTAINER_RT}/${CONTAINER_NAME}".  We actually already support using template strings for filename substitution.  We do have container name... we just don't have container runtime.  So, if we add that, then we could make the default be a template string that then doesn't change on a per-container basis.  yes, the value of the string is different after the substitution, but the default value is the same.

What to do for the parser is a little tougher.  Part of me thinks we should just have a different user-visible way for the customer to specify the parser, since it has a different meaning anyway.  Call it `default_parser` instead of `parser`.  We document `default_parser` as the parser to use *if* there is no parser specified for the container via an annotation.  That's different than `parser` which means "always use this value as the parser".

What do you think about those ideas?  Use `default_parser` and have a templated-value for the default rename_logfile.  We just would have to add in the runtime name as a templated variable.

Reply:
I think that makes sense.  I'll write some longer comments once I have made some changes.